## Part 1: Preprocessing

In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [ ]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [ ]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()


Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [ ]:
# Create a list of at least 10 column names to use as X data
# 'You can use any 10 columns you'd like' - Challenge 19 Instructions
X_list = ['Age', 'Education', 'JobInvolvement', 'JobSatisfaction', 'NumCompaniesWorked', 'PerformanceRating', 'TotalWorkingYears', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsWithCurrManager']
display(len(X_list))

# Create X_df using your selected columns
X_df = attrition_df[X_list]

# Show the data types for X_df
X_df.dtypes


10

Age                     int64
Education               int64
JobInvolvement          int64
JobSatisfaction         int64
NumCompaniesWorked      int64
PerformanceRating       int64
TotalWorkingYears       int64
YearsAtCompany          int64
YearsInCurrentRole      int64
YearsWithCurrManager    int64
dtype: object

In [ ]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=1)

In [ ]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

#There is no selected X data that needs converted into a numerical datatype. All chosen columns chosen were numerical

In [ ]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [ ]:
# Create a OneHotEncoder for the Department column
from sklearn.preprocessing import OneHotEncoder
ohedept = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# Fit the encoder to the training data
ohedept.fit(y_train[['Department']])

# Create two new variables by applying the encoder
# to the training and testing data
y_train_dept_encoded = ohedept.transform(y_train[['Department']])
y_test_dept_encoded = ohedept.transform(y_test[['Department']])


y_train_dept_encoded


array([[0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       ...,
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 0., 1.]])

In [ ]:
# Create a OneHotEncoder for the Attrition column
oheatt = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# Fit the encoder to the training data
oheatt.fit(y_train[['Attrition']])

# Create two new variables by applying the encoder
# to the training and testing data
y_train_att_encoded = oheatt.transform(y_train[['Attrition']])
y_test_att_encoded = oheatt.transform(y_test[['Attrition']])

y_train_att_encoded


array([[1., 0.],
       [1., 0.],
       [0., 1.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

## Create, Compile, and Train the Model

In [ ]:
# Find the number of columns in the X training data
num_columns = len(X_train.columns)

# Create the input layer
input_layer = layers.Input(shape=(num_columns,))

# Create at least two shared layers
shared_layer1 = layers.Dense(units=64, activation='relu')(input_layer)
shared_layer2 = layers.Dense(units=32, activation='relu')(shared_layer1)

In [ ]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
dept_hidden_layer = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
dept_output_layer = layers.Dense(3, activation='softmax', name='department_output')(dept_hidden_layer)


In [ ]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
att_hidden_layer = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
att_output_layer = layers.Dense(2, activation='sigmoid', name='attrition_output')(att_hidden_layer)


In [ ]:
# Create the model
model = Model(inputs=input_layer, outputs=[dept_output_layer, att_output_layer])

# Compile the model
model.compile(optimizer='adam',
              loss={'department_output': 'categorical_crossentropy', 'attrition_output': 'binary_crossentropy'},
              metrics={'department_output': 'accuracy', 'attrition_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)  │ (None, 10)             │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense (Dense)             │ (None, 64)             │            704 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_1 (Dense)           │ (None, 32)             │          2,080 │ dense[0][0]            │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_2 (Dense)           │ (None, 32)             │          1,056 │ dense_1[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_4 (Dense)           │ (None, 32)             │          1,056 │ dense_1[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             99 │ dense_2[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             66 │ dense_4[0][0]          │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 5,061 (19.77 KB)

 Trainable params: 5,061 (19.77 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Train the model
history = model.fit(X_train_scaled,
                    {'department_output': y_train_dept_encoded,
                     'attrition_output': y_train_att_encoded},
                    epochs=100)


Epoch 1/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - attrition_output_accuracy: 0.5580 - department_output_accuracy: 0.8629 - loss: 1.0522
Epoch 2/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8513 - department_output_accuracy: 0.8651 - loss: 0.8024
Epoch 3/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8283 - department_output_accuracy: 0.8853 - loss: 0.7453
Epoch 4/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8657 - department_output_accuracy: 0.8891 - loss: 0.6863
Epoch 5/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8479 - department_output_accuracy: 0.8915 - loss: 0.6959
Epoch 6/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8629 - department_output_accuracy: 0.8932 - loss: 0.6442
Epoch 7/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8513 - department_output_accuracy: 0.8876 - loss: 0.6641
Epoch 8/100
35/35 ━━━━━━━━━

In [ ]:
# Evaluate the model with the testing data
test_accuracy = model.evaluate(X_test_scaled,
              {'department_output': y_test_dept_encoded,
              'attrition_output': y_test_att_encoded},
              verbose=2)

12/12 - 1s - 65ms/step - attrition_output_accuracy: 0.7527 - department_output_accuracy: 0.5598 - loss: 3.4103


In [ ]:
# Print the accuracy for both department and attrition
print(f'Department Predictions Accuracy: {test_accuracy[2]}')
print(f'Attrition Predictions Accuracy: {test_accuracy[1]}')

Department Predictions Accuracy: 0.5597826242446899
Attrition Predictions Accuracy: 0.7527173757553101


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. When analyzing the results achieved in the previous cell, it could be determined that accuracy might not. be the best metric to use on this data. I bring this up due to how *low* the accuracy results are on the prediction side of things, like how the department prediction accuracy is at *0.559 or 0.56*. The predicted attrition accuracy is significantly higher by a little over *25%* at *0.752 or 0.75*, however, this is still *low* by accuracy standards.
2. At first, I held confusion as to what the different activations were. I had originally thought both would be 'softmax,'  however, after looking into more explanations I determined I had made an error. The only output layer that had 'softmax' as an activation was the *Department* output layer, as the original dataframe displayed the company having three different departments, three different classes... the department column having multiple classes in it determines it is multiclass, hence why 'softmax' was used. After the initial error I had made with thinking both output layers utilized 'softmax,' I had come to realize that the 'Attrition' column simply held yes and no content, '1's and '0's. Since it utilized only '1's and '0's (and two columns when encoded), it is still a sigmoid activation, as it can only result in *[1,0]* and *[0,1]* respectively.
3. When thinking of how to possibly improve my model, only a few ideas come to mind. My first thought would be to tune the hyperparameters. Using either Cross-Validation or Grid/Random search I could find the 10 best parameters to build my model with instead of choosing what I thought would be the best with no external context. Another possible improvement can comee from garnering more domain knowledge, this can also assist me and inform me of bettere feeature selection, engineering, and model interpretation; it can help me make better insights and improvements to thee model. Regularization techniques could possibly assist in the creation of a better model, especially since this dataset collection has 27 different features (which can be considered as a lot). Lastly, I believe that model evaluation could help, moreso in the usage of appropriate metrics side of things; as I noted previously, I don't believe that accuracy was the best metric that could have been used in this model, thinking this through more, I believe that precision could have been a better metric to utilize overall.